### ISSUE TITLE
Enable hierarchal namespaces governed by smart contract

### ISSUE THREAD

**Issue by larrysalibra** [bnswg/initiatives#4]
BNS as it exists today has three types of names: 

* namespaces - these look like top-level domains in the legacy DNS system and are the on-chain registration of defined prices and expiration rules
* names - these are strings registered on-chain in certain namespace
* subdomains - these are strings registered by the owner of a certain name and stored offchain. They are not subject to the price or expiration rules of names.

@MarvinJanssen suggested that instead of the simple pricing and expiration rules of BNSv2, that each namespace could be instead governed by its own smart contract. Namespace smart contracts could have their own rules for how names (or sub-namespaces) in their namespace are governed. It is easy to imagine creative ways that this could be used to unlock new opportunities for our ecosystem and seems like something worth exploring.

**Comment by friedger** [comment]
Subdomains are stored on-chain in bulk txs. Owners of subdomains have the same on-chain experience as name owners.

Subdomains are subject to price and expiration rules of the owner of the name. Subdomain owners trust the owner of the name to handle name transactions in their name.

**Comment by markmhendrickson** [comment]
> Owners of subdomains have the same on-chain experience as name owners

This is true except (as we discuss briefly yesterday) owners of subdomains can't transfer them to another address unless they get permission top-level name owner somehow?

**Comment by larrysalibra** [comment]
> Subdomains are stored on-chain in bulk txs. Owners of subdomains have the same on-chain experience as name owners.

@friedger Where are subdomains stored on-chain? In the[ Stacks docs](https://docs.stacks.co/build-apps/references/bns), they say that subdomains are stored off-chain. Is it possible to resolve subdomains with the BNS smart contract?
<img width="585" alt="Screen Shot 2022-02-25 at 20 16 16" src="https://user-images.githubusercontent.com/597182/155713795-8f687491-eb51-4530-af8d-08761c592a9c.png">

**Comment by friedger** [comment]
@larrysalibra They are stored in attachments of an name-update tx of the registrar's on-chain name.

It is not possible to look them up via BNS smart contract.

**Comment by friedger** [comment]
Ryder Handles (https://handles.ryder.id) is an implementation of namespaces governed by smart contract: https://github.com/Light-Labs/stacks-name-system
